### PR TITLE
handbook

### PR DESCRIPTION
**Pull request by michaevelli**

### CODE REVIEW

**Comment by michaevelli** [comment]
"Revert "Prettied up errors" shouldn't have been pushed help

File: handlers/handbook.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+		}
+		if cond == "" {
+			if strings.Contains(line, "Prerequisite") {
+				cond = strings.Split(strings.Split(string(line), ">")[1], "<")[0]
+			}
+		}
+		if title != "" && desc != "" && term != "" && cond != "" {
+			// stop early if all fields are filled
+			break
+		}
```
Comment (VineBalloon) [inline handlers/handbook.go]: This is super-hacky, I like it.

File: handlers/handbook.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	// create and send message
+	message := commands.NewSend(msg.ChannelID)
+	embed := makeMessage(url, title, desc, term, cond)
+	return message.Embed(embed), nil
+}
+
+
+func getResponse(Code string, Graduate string) (string, []byte, int, error) {
```
Comment (VineBalloon) [inline handlers/handbook.go]: In this case I would either write a function annotation or use named return values to make the function's... functionality more clear.

File: handlers/tags.go [hunk 3]
```diff
@@ -46,6 +46,10 @@ var (
 	ErrNoUser = errors.New("you don't have a tag on this platform")
 	// ErrUserNotFound means the user queried a username that doesn't exist on the server
 	ErrUserNotFound = errors.New("user not found")
+	// ErrAddSpam means the user tried to add while a new platform was being waited on
+	ErrAddSpam = errors.New("please do not try add anything while I'm waiting")
+	// ErrCleanSpam means the user tried to clean while a clean is in progress
+	ErrCleanSpam = errors.New("already cleaning, please be patient")
```
Comment (VineBalloon) [inline handlers/tags.go]: Thanks!

File: handlers/handbook.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
+package handlers
+
+import (
+	"errors"
```
[the remark this answers is not in the document]
Comment (michaevelli) [reply handlers/handbook.go]: Done

File: handlers/handbook.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	"strings"
+	"bytes"
+	"regexp"
+	"net/http"
+	"html"
+	"io/ioutil"
+	"github.com/bwmarrin/discordgo"
+	"github.com/unswpcsoc/PCSocBot/commands"
```
[the remark this answers is not in the document]
Comment (michaevelli) [reply handlers/handbook.go]: Ok

File: handlers/handbook.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+)
+
+var (
+
```
[the remark this answers is not in the document]
Comment (michaevelli) [reply handlers/handbook.go]: Done

File: handlers/handbook.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+		return nil, err
+	}
+
+	if status != 200 {
+		// probably a 404 error but whatever it is, the page could not be accessed
+		return nil, ErrNotFound
+	}
+
+	// initialise info
+	title, desc, term, cond := "", "", "", ""
```
[the remark this answers is not in the document]
Comment (michaevelli) [reply handlers/handbook.go]: Done

File: handlers/handbook.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+	match, err := regexp.MatchString(`^[A-Za-z]{4}[0-9]{4}$`, h.Code)
+	if err != nil {
+		return nil, err
+	}
+	if !match {
+		return nil, ErrInvalidFormat
+	}
+
+	//assume undergraduate
+	url, htmltext, status, err := getResponse(h.Code, "undergraduate")
```
[the remark this answers is not in the document]
Comment (michaevelli) [reply handlers/handbook.go]: Done

File: handlers/handbook.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+		return nil, ErrNotFound
+	}
+
+	// initialise info
+	title, desc, term, cond := "", "", "", ""
+
+	// extract info via shifty means
+	lines := bytes.Split(htmltext, []byte("\n"))
+	for i, line := range lines {
+		line := string(line[:])
```
[the remark this answers is not in the document]
Comment (michaevelli) [reply handlers/handbook.go]: `line` as declared in the for statement is a slice of bytes. Within the for loop, I never refer to the data as a slice of bytes, only as a string, so I redeclared `line` in the first line of the for loop as a string.

File: handlers/handbook.go (hunk 2 above)
Comment (michaevelli) [reply handlers/handbook.go]: Done

File: handlers/tags.go (hunk 3 above)
Comment (michaevelli) [reply handlers/tags.go]: I wasn't supposed to touch this

File: handlers/handbook.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (VineBalloon) [reply handlers/handbook.go]: Oh I see, so you're doing an explicit cast to string, makes sense. I fail to see the need for the `[:]` slicing, could you clarify? Also I'd use a different name for the line or even the loop's range value e.g. `lineBytes` or `lineString`.

File: handlers/tags.go (hunk 3 above)
Comment (VineBalloon) [reply handlers/tags.go]: Ok

File: handlers/handbook.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (michaevelli) [reply handlers/handbook.go]: `line` inside loop is now `lineString`, `[:]` has been removed (was there from not reading stackoverflow q/a properly)

**Pull request merged**